.doom.d/config.org
#+TITLE: pgrogan Emacs Configuration
* Base Configuration
** Personal Information
Basic user information
#+BEGIN_SRC emacs-lisp :tangle config.el
(setq user-full-name "Patrick Grogan"
      user-mail-address "pgrogan@gmail.com")
#+End_SRC
** Font and Size
*** Font face
 #+BEGIN_SRC emacs-lisp :tangle config.el
 (setq doom-font (font-spec :family "FiraCode Nerd Font" :size 16)
        doom-unicode-font (font-spec :name "Noto Color Emoji" :size 20)
        doom-variable-pitch-font (font-spec :family "Ubuntu" :size 16))
#+END_SRC

#+RESULTS:
: #<font-spec nil nil Ubuntu nil nil nil nil nil 16 nil nil nil nil>

*** Emojis
 #+BEGIN_SRC emacs-lisp :tangle packages.el
 (package! emojify)
 #+END_SRC
** Whitespace management
 Some of these settings are the doom's default but I just want to make sure
 #+BEGIN_SRC emacs-lisp :tangle config.el
     (setq-default indent-tabs-mode nil)
     (setq-default tab-width 4)
     (setq-default standard-indent 4)
     (setq-default tab-stop-list (number-sequence 4 120 4))
 #+END_SRC
 Also make sure we strip all whitespaces on save
 #+BEGIN_SRC emacs-lisp :tangle config.el
     (add-hook 'before-save-hook 'delete-trailing-whitespace)
 #+END_SRC
** Line numbers
 I really like to have both absolute and relative line numbers.
 #+BEGIN_SRC emacs-lisp :tangle config.el
 (setq display-line-numbers-type 'relative)
 #+END_SRC
** Word delimiter
 For all programming language, I prefere that the _ character is
 always part of the word. This is the Vim way
 #+BEGIN_SRC emacs-lisp :tangle config.el
 (add-hook! prog-mode-hook (modify-syntax-entry ?_ "w"))
 #+END_SRC
** Load Theme
#+BEGIN_SRC emacs-lisp :tangle config.el
 (load-theme 'doom-dark+ t)
;(def-package! doom-modeline
;  :defer t
;  :hook (after-init . doom-modeline-init))
 (after! doom-modeline
   (setq doom-modeline-major-mode-color-icon t))
 #+END_SRC
** Undo
We want to use undo-fu as a undo manager.
#+BEGIN_SRC emacs-lisp :tangle packages.el
(package! undo-fu)
#+END_SRC
#+BEGIN_SRC emacs-lisp :tangle config.el
(evil-set-undo-system 'undo-fu)
#+END_SRC

* Filetype mapping
** Makefiles
 All sort of makefiles should be mapped in makefile-mode. For example, make_reg or Makefile.mentor
  #+BEGIN_SRC emacs-lisp :tangle config.el
  (add-to-list 'auto-mode-alist '("make_\.\*" . makefile-mode))
  (add-to-list 'auto-mode-alist '("Makefile\.\*" . makefile-mode))
  #+END_SRC
* Custom functions
** Display ansi colors in file
 Some time, due to log checking, it happens I want to display the ansi color
 instead of only code. This function helps me with that.
 #+BEGIN_SRC emacs-lisp :tangle packages.el
 (package! ansi-color)
 #+END_SRC

 #+BEGIN_SRC emacs-lisp :tangle config.el
 (after! ansi-color
   (defun display-ansi-colors ()
     (interactive)
     (ansi-color-apply-on-region (point-min) (point-max))))
 #+END_SRC
* Projectile Configuration
 1- Set the indexing method to alien
 2- Set the search path to ~/workspace
 3- Defining a custom switching function that workd with lsp
 4- Hook that function to projectile project switch
 #+BEGIN_SRC emacs-lisp :tangle packages.el
 (package! virtualenvwrapper)
 #+END_SRC
 #+BEGIN_SRC emacs-lisp :tangle config.el
 (after! projectile
   (setenv "WORKON_HOME" "~/virtualenvs")
   (setq projectile-indexing-mode 'alien
         projectile-project-search-path '("~/workspace"))
   (defun setup_env (&optional name)
     (interactive
      (list
       (completing-read "Work on: " (pyvenv-virtualenv-list)
                        nil t nil 'pyenv-workon-history nil nil)))
     (pyvenv-deactivate)
     (setenv "TOOLS_PATH" (concat (projectile-project-root) "/tools"))
     (setenv "PYTHONPATH" (concat (projectile-project-root) ":" (getenv "TOOLS_PATH") "/cocotb:" (getenv "TOOLS_PATH") "/themis_fw:"))
     ;(pyvenv-virtualenv-list)
    (pyvenv-workon name)
    (lsp)
    (setq projectile-tags-command (concat (projectile-project-root)"scripts/etags/verilog_etags " (projectile-project-root) "rtl"))
    (setq projectile-tags-file-name (concat (projectile-project-root) "rtl/TAGS"))
    )
  (add-hook! 'projectile-after-switch-project-hook #'setup_env))
#+END_SRC

* LSP Configuration
 #+BEGIN_SRC emacs-lisp :tangle config.el
(after! lsp-ui
  lsp-ui-doc-enable t
  lsp-ui-doc-mode t)
(map! :leader
      :after lsp-ui
      :desc "Jump backward"
      "c ," #'lsp-ui-peek-jump-backward)
(map! :leader
      :after lsp-ui
      :desc "Jump backward"
      "c ." #'lsp-ui-peek-jump-forward)
 #+END_SRC
* Python specific configuration
 #+BEGIN_SRC emacs-lisp :tangle config.el
 (after! flycheck
   ;(flycheck-add-next-checker 'python-pylint 'python-flake8)
   (add-hook! 'flycheck-mode-hook
     (defun set-python-flycheck ()
       (when (eq major-mode 'python-mode)
         (setq flycheck-checker 'python-flake8)))))
         ;; This will re-enable pylint
         ;(flycheck-disable-checker 'python-pylint t)
 #+END_SRC
#+END_SRC
* Magit Configuration

#+RESULTS:
| +emacs-lisp-reduce-flycheck-errors-in-emacs-config-h | set-python-flycheck | flycheck-mode-set-explicitly | doom-modeline-update-flycheck-text | doom-modeline-update-flycheck-icon | +syntax-init-popups-h |

* Additional packages
** Graphviz dot mode
(package! graphviz-dot-mode)
  #+END_SRC
** LUA Mode
  #+BEGIN_SRC emacs-lisp :tangle packages.el
  (package! lua-mode)
  #+END_SRC
** Python
  We use jedi as a company backend.
  #+BEGIN_SRC emacs-lisp :tangle packages.el
  (package! company-jedi)
  #+END_SRC
  #+BEGIN_SRC emacs-lisp :tangle config.el
  (use-package! company-jedi
    :config
    (add-to-list 'company-backends 'company-jedi)
    (setq jedi:complete-on-dot t)
    :hook
    (inferior-python-mode . jedi:setup)
    (python-mode . jedi:setup))
  #+END_SRC

** Jinja2
  #+BEGIN_SRC emacs-lisp :tangle packages.el
  (package! jinja2-mode)
  #+END_SRC
  #+BEGIN_SRC emacs-lisp :tangle config.el
  (add-to-list 'auto-mode-alist '("\\.jinja2?\\'" . jinja2-mode))
  #+END_SRC
** Verilog Mode
  #+BEGIN_SRC emacs-lisp :tangle packages.el
  (package! verilog-mode)
  #+END_SRC
  #+BEGIN_SRC emacs-lisp :tangle config.el
  (after! verilog-mode
    (setq verilog-auto-newline nil
          verilog-tab-auto-indent nil
          verilog-case-indent 4
          verilog-cexp-indent 4
          verilog-highlight-grouping-keyword t
          verilog-highlight-modules nil
          verilog-indent-level 4
          verilog-indent-level-behavioral 4
          verilog-indent-level-declaration 4
          verilog-indent-level-module 4
          verilog-auto-lineup 'assignment)
    (define-key verilog-mode-map (kbd ";") 'self-insert-command)
    (define-key verilog-mode-map (kbd ":") 'self-insert-command)
    (define-key verilog-mode-map (kbd "RET") 'evil-ret)
    (define-key verilog-mode-map (kbd "TAB") 'tab-to-tab-stop)
    ; Load verilog mode only when needed
    (autoload 'verilog-mode "verilog-mode" "Verilog mode" t)
    ; Any files that ends in .v, .dv or .sv should be in verilog mode
    (add-to-list 'auto-mode-alist '("\\.[ds]?vh?\\'" . verilog-mode))
    ; Any files in verilog mode should have their keywords colorized
    (lsp-register-client
     (make-lsp-client :new-connection (lsp-stdio-connection '("svls"))
                      :major-modes '(verilog-mode)
                      :priority: -1))
    (add-hook 'verilog-mode-hook '(lambda()
        (font-lock-mode 1)
        (lsp)
        (flycheck-mode t)
        (add-to-list 'lsp-language-id-configuration '(verilog-mode . "verilog")))))
  #+END_SRC
** Yaml
  #+BEGIN_SRC emacs-lisp :tangle packages.el
  (package! yaml-mode)
  #+END_SRC
** Vue mode
  #+Begin_SRC emacs-lisp :tangle packages.el
  (package! vue-mode)
  #+END_SRC
  #+BEGIN_SRC emacs-lisp :tangle config.el
  (after! vue-mode
    (add-hook 'vue-mode-hook #'lsp))
  #+END_SRC
* Org Mode Configuration
#+BEGIN_SRC emacs-lisp :tangle packages.el
(package! org-bullets)
#+END_SRC
** Org Bullets configuration
First we configure org bullets.
#+BEGIN_SRC emacs-lisp :tangle config.el
(use-package! org-bullets
  :after org
  :hook (org-mode . org-bullets-mode)
  :config
  (setq org-bullets-bullet-list '("○" "☉" "◎" "◉" "○" "◌" "◎" "●" "◦" "◯")))
#+END_SRC
** Fonts configuration
#+BEGIN_SRC emacs-lisp :tangle packages.el
(package! mixed-pitch)
#+END_SRC
#+BEGIN_SRC emacs-lisp :tangle config.el
(add-hook 'text-mode-hook 'mixed-pitch-mode)
(defun nolinum ()
  (interactive)
  (setq doom--line-number-style nil)
  (setq display-line-numbers nil)
)
(add-hook 'org-mode-hook 'nolinum)
(after! org
  (set-face-attribute 'org-table nil :inherit 'fixed-pitch)
  (set-face-attribute 'org-code nil :inherit 'fixed-pitch)
  (set-face-attribute 'org-block nil :inherit 'fixed-pitch))
#+END_SRC


.doom.d/config.el
(setq user-full-name "Patrick Grogan"
      user-mail-address "pgrogan@gmail.com")

 (setq doom-font (font-spec :family "FiraCode Nerd Font" :size 16)
        doom-unicode-font (font-spec :name "Noto Color Emoji" :size 20)
        doom-variable-pitch-font (font-spec :family "Helvetica Neue" :size 16))

     (setq-default indent-tabs-mode nil)
     (setq-default tab-width 4)
     (setq-default standard-indent 4)
     (setq-default tab-stop-list (number-sequence 4 120 4))

     (add-hook 'before-save-hook 'delete-trailing-whitespace)

 (setq display-line-numbers-type 'relative)

 (add-hook! prog-mode-hook (modify-syntax-entry ?_ "w"))

 (load-theme 'doom-dark+ t)

  (add-to-list 'auto-mode-alist '("make_\.\*" . makefile-mode))
  (add-to-list 'auto-mode-alist '("Makefile\.\*" . makefile-mode))

 (after! ansi-color
   (defun display-ansi-colors ()
     (interactive)
     (ansi-color-apply-on-region (point-min) (point-max))))

 (after! projectile
   (setenv "WORKON_HOME" "~/virtualenvs")
   (setq projectile-indexing-mode 'alien
         projectile-project-search-path '("~/workspace"))
   (defun setup_env (&optional name)
     (interactive
      (list
       (completing-read "Work on: " (pyvenv-virtualenv-list)
                        nil t nil 'pyenv-workon-history nil nil)))
     (pyvenv-deactivate)
     (setenv "TOOLS_PATH" (concat (projectile-project-root) "/tools"))
     (setenv "PYTHONPATH" (concat (projectile-project-root) ":" (getenv "TOOLS_PATH") "/cocotb:" (getenv "TOOLS_PATH") "/themis_fw:"))
     ;(pyvenv-virtualenv-list)

     (pyvenv-workon name)
     (lsp)
     (setq projectile-tags-command (concat (projectile-project-root)"scripts/etags/verilog_etags " (projectile-project-root) "rtl"))
     (setq projectile-tags-file-name (concat (projectile-project-root) "rtl/TAGS"))
     )
   (add-hook! 'projectile-after-switch-project-hook #'setup_env))

(after! lsp-ui
  lsp-ui-doc-enable t
  lsp-ui-doc-mode t)
(map! :leader
      :after lsp-ui
      :desc "Jump backward"
      "c ," #'lsp-ui-peek-jump-backward)
(map! :leader
      :after lsp-ui
      :desc "Jump backward"
      "c ." #'lsp-ui-peek-jump-forward)

 (after! flycheck
   ;(flycheck-add-next-checker 'python-pylint 'python-flake8)
   (add-hook! 'flycheck-mode-hook
     (defun set-python-flycheck ()
       (when (eq major-mode 'python-mode)
         (setq flycheck-checker 'python-flake8)))))
         ;; This will re-enable pylint
         ;(flycheck-disable-checker 'python-pylint t)

  (use-package! company-jedi
    :config
    (add-to-list 'company-backends 'company-jedi)
    (setq jedi:complete-on-dot t)
    :hook
    (inferior-python-mode . jedi:setup)
    (python-mode . jedi:setup))

  (add-to-list 'auto-mode-alist '("\\.jinja2?\\'" . jinja2-mode))

  (after! verilog-mode
    (setq verilog-auto-newline nil
          verilog-tab-auto-indent nil
          verilog-case-indent 4
          verilog-cexp-indent 4
          verilog-highlight-grouping-keyword t
          verilog-highlight-modules nil
          verilog-indent-level 4
          verilog-indent-level-behavioral 4
          verilog-indent-level-declaration 4
          verilog-indent-level-module 4
          verilog-auto-lineup 'assignment)
    (define-key verilog-mode-map (kbd ";") 'self-insert-command)
    (define-key verilog-mode-map (kbd ":") 'self-insert-command)
    (define-key verilog-mode-map (kbd "RET") 'evil-ret)
    (define-key verilog-mode-map (kbd "TAB") 'tab-to-tab-stop)
    ; Load verilog mode only when needed
    (autoload 'verilog-mode "verilog-mode" "Verilog mode" t)
    ; Any files that ends in .v, .dv or .sv should be in verilog mode
    (add-to-list 'auto-mode-alist '("\\.[ds]?vh?\\'" . verilog-mode))
    ; Any files in verilog mode should have their keywords colorized
    (add-hook 'verilog-mode-hook '(lambda () (font-lock-mode 1))))

  (after! vue-mode
    (add-hook 'vue-mode-hook #'lsp))

(use-package! org-bullets
  :hook (org-mode . org-bullets-mode)
  :config
  (setq org-bullets-bullet-list '("○" "☉" "◎" "◉" "○" "◌" "◎" "●" "◦" "◯")))

(add-hook 'text-mode-hook 'mixed-pitch-mode)
(defun nolinum ()
  (interactive)
  (setq doom--line-number-style nil)
  (setq display-line-numbers nil)
)
(add-hook 'org-mode-hook 'nolinum)
(after! org
  (set-face-attribute 'org-table nil :inherit 'fixed-pitch)
  (set-face-attribute 'org-code nil :inherit 'fixed-pitch)
  (set-face-attribute 'org-block nil :inherit 'fixed-pitch))
